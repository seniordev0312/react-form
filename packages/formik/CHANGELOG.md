# formik

## 2.2.3

### Patch Changes

- [`e0a28e6`](https://github.com/formium/formik/commit/e0a28e6872ebfd06e636aac84829b60d704b0694) [#2867](https://github.com/formium/formik/pull/2867) Thanks [@umidbekkarimov](https://github.com/umidbekkarimov)! - Fix low priority validation for browser password autofill.

* [`958d67c`](https://github.com/formium/formik/commit/958d67ca2c3e006031c31150ea0a42248b28ffc7) [#2874](https://github.com/formium/formik/pull/2874) Thanks [@maddhruv](https://github.com/maddhruv)! - fix FastField initial value when input type is radio or checkbox

## 2.2.2

### Patch Changes

- [`00f95ec`](https://github.com/formium/formik/commit/00f95ec4ec5266eed8ad4e97b76321205c704d51) [#2854](https://github.com/formium/formik/pull/2854) Thanks [@umidbekkarimov](https://github.com/umidbekkarimov)! - Fix low priority validation race condition.

## 2.2.1

### Patch Changes

- [`e04886d`](https://github.com/formium/formik/commit/e04886db15c7e9b96516b4bd5a1b89d0e895bb7d) [#2820](https://github.com/formium/formik/pull/2820) Thanks [@wellyshen](https://github.com/wellyshen)! - Fixed bug with scheduler and `validateFormWithLowPriority` method not be scheduled correctly

## 2.2.0

### Minor Changes

- [`4148181`](https://github.com/formium/formik/commit/41481819f9187de79c4d948aeaa4ca1d33c53ed7) [#2794](https://github.com/formium/formik/pull/2794) Thanks [@jaredpalmer](https://github.com/jaredpalmer)! - `setValue` can now optionally accept a function as a callback, exposing `React.SetStateAction` functionality. Previously, only the entire object was
  allowed which caused issues with stale props.

  ```tsx
  setValues(prevValues => ({...prevValues, ... }))
  ```

## 2.1.7

### Patch Changes

- [`bda9f41`](https://github.com/formium/formik/commit/bda9f41931fac382eec26b4f1283b881b6bbc240) [#2785](https://github.com/formium/formik/pull/2785) Thanks [@jaredpalmer](https://github.com/jaredpalmer)! - Replace all instances of the deprecated React.SFC type with React.FC
