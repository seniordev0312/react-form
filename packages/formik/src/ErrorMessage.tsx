import * as React from 'react';
import { useFieldError, useFieldTouched } from './useField';
import { isFunction } from './utils';
export interface ErrorMessageProps {
  name: string;
  className?: string;
  component?: string | React.ComponentType;
  children?: (errorMessage: string) => React.ReactNode;
  render?: (errorMessage: string) => React.ReactNode;
}
export function ErrorMessage({
  component,
  render,
  children,
  name,
  ...rest
}: ErrorMessageProps) {
  const touch = useFieldTouched(name);
  const error = useFieldError(name);
  return !!touch && !!error
    ? render
      ? isFunction(render)
        ? render(error)
        : null
      : children
      ? isFunction(children)
        ? children(error)
        : null
      : component
      ? React.createElement(component, rest as any, error)
      : error
    : null;
}
