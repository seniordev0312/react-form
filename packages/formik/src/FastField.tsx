import * as React from 'react';
import invariant from 'tiny-warning';
import {
  FieldInputProps,
  FieldMetaProps,
  GenericFieldHTMLAttributes,
  SharedFieldProps,
} from './types';
import { useField, UseFieldProps } from './useField';
import { isEmptyChildren, isFunction } from './utils';

export type FastFieldProps<FieldValue = any> = {
  field: FieldInputProps<FieldValue>;
  meta: FieldMetaProps<FieldValue>;
};

export type FastFieldConfig<V = any> = UseFieldProps<V> &
  SharedFieldProps<FastFieldProps<V>> & {
    /**
     * Override FastField's default shouldComponentUpdate
     * @deprecated
     */
    shouldUpdate?: (nextProps: any, props: {}) => boolean;
  };

export function FastField<V = any>({
  render,
  children,
  as: is, // `as` is reserved in typescript lol
  component,
  shouldUpdate,
  ...props
}: GenericFieldHTMLAttributes & FastFieldConfig<V>) {
  React.useEffect(() => {
    invariant(
      !render,
      `<FastField render> has been deprecated. Please use a child callback function instead: <FastField name={${name}}>{props => ...}</FastField> instead.`
    );
    invariant(
      !(component && render),
      'You should not use <FastField component> and <FastField render> in the same <FastField> component; <FastField component> will be ignored'
    );

    invariant(
      !(is && children && isFunction(children)),
      'You should not use <FastField as> and <FastField children> as a function in the same <FastField> component; <FastField as> will be ignored.'
    );

    invariant(
      !(component && children && isFunction(children)),
      'You should not use <FastField component> and <FastField children> as a function in the same <FastField> component; <FastField component> will be ignored.'
    );

    invariant(
      !(render && children && !isEmptyChildren(children)),
      'You should not use <FastField render> and <FastField children> in the same <FastField> component; <FastField children> will be ignored'
    );
    invariant(
      !shouldUpdate,
      `<FastField shouldUpdate> has been deprecated. Please create a custom component with useField and React.memo() instead.`
    );
    // eslint-disable-next-line
  }, []);

  const [field, meta] = useField<V>(props);

  if (render) {
    return render({ field, meta });
  }

  if (isFunction(children)) {
    return children({ field, meta });
  }

  const { innerRef, parse, format, formatOnBlur, validate, ...rest } = props;

  if (component) {
    // This behavior is backwards compat with earlier Formik 0.9 to 1.x
    if (typeof component === 'string') {
      return React.createElement(
        component,
        { ref: innerRef, ...field, ...rest },
        children
      );
    }
    return React.createElement(component, { field, meta, ...rest }, children);
  }

  // default to input here so we can check for both `as` and `children` above
  const asElement = is || 'input';

  if (typeof asElement === 'string') {
    return React.createElement(
      asElement,
      { ref: innerRef, ...field, ...rest },
      children
    );
  }
  return React.createElement(
    asElement as React.FunctionComponent<FieldInputProps<V>>,
    { ...field, ...rest },
    children
  );
}
