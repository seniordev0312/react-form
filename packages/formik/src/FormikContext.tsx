import * as React from 'react';
import { FormikContextType } from './types';
import {
  createContext,
  useContext,
  useContextSelector,
} from 'use-context-selector';
import invariant from 'tiny-warning';

export const FormikContext = createContext<FormikContextType<any>>(
  undefined as any
);

export const FormikProvider = FormikContext.Provider;

export function useFormikContext<Values>() {
  return useContext<FormikContextType<Values>>(FormikContext);
}

export function FormikConsumer<Values = any>({
  children,
}: {
  children: (formik: FormikContextType<Values>) => React.ReactNode;
}) {
  const formik = useFormikContext<Values>();

  invariant(
    !!formik,
    `Formik context is undefined, please verify you are calling useFormikContext() as child of a <Formik> component.`
  );

  return <>{children(formik)}</>;
}

export function useFormikContextSelector<Values = any, Slice = any>(
  selector: (value: FormikContextType<Values>) => Slice
): Slice {
  return useContextSelector(FormikContext, selector);
}
