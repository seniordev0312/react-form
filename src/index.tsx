import * as React from 'react';
const hoistNonReactStatics = require('hoist-non-react-statics');

/**
 * Transform Yup ValidationError to a more usable object
 */
export function yupToFormErrors(yupError: any): FormikErrors {
  let errors: FormikErrors = {};
  for (let err of yupError.inner) {
    if (!errors[err.path]) {
      errors[err.path] = err.message;
    }
  }
  return errors;
}

/**
 * Given an object, map keys to boolean
 */
export function touchAllFields<T>(fields: T): { [field: string]: boolean } {
  let touched: { [field: string]: boolean } = {};
  for (let k of Object.keys(fields)) {
    touched[k] = true;
  }
  return touched;
}
/**
 * Validate a yup schema.
 */
export function validateFormData<T>(data: T, schema: any): Promise<void> {
  let validateData: any = {};
  for (let k in data) {
    if (data.hasOwnProperty(k)) {
      const key = String(k);
      validateData[key] = (data as any)[key] !== ''
        ? (data as any)[key]
        : undefined;
    }
  }
  return schema.validate(validateData, { abortEarly: false });
}

export interface FormikValues {
  [field: string]: any;
}

// @todo make this limited to keys of values
export interface FormikErrors {
  [field: string]: string;
}

// @todo make this limited to keys of values
// interfact FormikTouched<Values, Field extends keyof Values> ??
export interface FormikTouched {
  [field: string]: boolean;
}

/**
 * Formik configuration options
 */
export interface FormikConfig<Props, Values, Payload> {
  displayName?: string;
  /* Map props to the form values */
  mapPropsToValues?: (props: Props) => Values;
  /* Map form values to submission payload */
  mapValuesToPayload?: (values: Values) => Payload;
  /*  Yup Schema */
  validationSchema: any;
  /* Submission handler */
  handleSubmit: (payload: Payload, formikBag: FormikBag<Props, Values>) => void;
}

/**
 * Formik state tree
 */
export interface FormikState<V> {
  /* Form values */
  values: V;
  /* Top level error, in case you need it */
  error?: any;
  /** map of field names to specific error for that field */
  errors: FormikErrors;
  /** map of field names to whether the field has been touched */
  touched: FormikTouched;
  /** whether the form is currently submitting */
  isSubmitting: boolean;
}

/**
 * Formik state helpers
 */
export interface FormikActions<P, V> {
  /* Manually set top level error */
  setError: (e: any) => void;
  /* Manually set Errors */
  setErrors: (errors: FormikErrors) => void;
  /* Manually set isSubmitting */
  setSubmitting: (isSubmitting: boolean) => void;
  /* Manually set touched fields */
  setTouched: (touched: FormikTouched) => void;
  /* Manually set values  */
  setValues: (values: V) => void;
  /* Reset form */
  resetForm: (nextProps?: P) => void;
}

/**
 * Formik form event handlers 
 */
export interface FormikHandlers {
  /* Form submit handler */
  handleSubmit: (e: React.FormEvent<HTMLFormElement>) => void;
  /* Classic React change handler, keyed by input name */
  handleChange: (e: React.ChangeEvent<any>) => void;
  /* Mark input as touched */
  handleBlur: (e: any) => void;
  /* Change value of form field directly */
  handleChangeValue: (name: string, value: any) => void;
  /* Reset form event handler  */
  handleReset: () => void;
}

/**
 * State, handlers, and helpers injected as props into the wrapped form component.
 */
export type InjectedFormikProps<Props, Values> = Props &
  FormikState<Values> &
  FormikActions<Props, Values> &
  FormikHandlers;

/**
 * Formik actions + { props }
 */
export type FormikBag<P, V> = { props: P } & FormikActions<P, V>;

export type CompositeComponent<P> =
  | React.ComponentClass<P>
  | React.StatelessComponent<P>;

export interface ComponentDecorator<TOwnProps, TMergedProps> {
  (component: CompositeComponent<TMergedProps>): React.ComponentClass<
    TOwnProps
  >;
}

export interface InferableComponentDecorator<TOwnProps> {
  <T extends CompositeComponent<TOwnProps>>(component: T): T;
}

export default function formik<Props, Values extends FormikValues, Payload>({
  displayName,
  mapPropsToValues = vanillaProps => {
    let values: Values = {} as Values;
    for (let k in vanillaProps) {
      if (
        vanillaProps.hasOwnProperty(k) &&
        typeof vanillaProps[k] !== 'function'
      ) {
        values[k] = vanillaProps[k];
      }
    }
    return values;
  },
  mapValuesToPayload = values => {
    // in this case Values and Payload are the same.
    const payload = (values as any) as Payload;
    return payload;
  },
  validationSchema,
  handleSubmit,
}: FormikConfig<Props, Values, Payload>): ComponentDecorator<
  Props,
  InjectedFormikProps<Props, Values>
> {
  return function wrapWithFormik(
    WrappedComponent: CompositeComponent<InjectedFormikProps<Props, Values>>
  ): any {
    class Formik extends React.Component<Props, FormikState<Values>> {
      public static displayName = `Formik(${displayName ||
        WrappedComponent.displayName ||
        WrappedComponent.name ||
        'Component'})`;
      public static WrappedComponent = WrappedComponent;
      public props: Props;

      constructor(props: Props) {
        super(props);
        this.state = {
          values: mapPropsToValues(props),
          errors: {},
          touched: {},
          isSubmitting: false,
        };
      }

      setErrors = (errors: FormikErrors) => {
        this.setState({ errors });
      };

      setTouched = (touched: FormikTouched) => {
        this.setState({ touched });
      };

      setValues = (values: Values) => {
        this.setState({ values });
      };

      setError = (error: any) => {
        this.setState({ error });
      };

      setSubmitting = (isSubmitting: boolean) => {
        this.setState({ isSubmitting });
      };

      handleChange = (e: React.ChangeEvent<any>) => {
        e.persist();
        const { type, name, id, value, checked } = e.target;
        const field = name ? name : id;
        const val = /number|range/.test(type)
          ? parseFloat(value)
          : /checkbox/.test(type) ? checked : value;

        const { values } = this.state;
        // Set form fields by name
        this.setState(state => ({
          ...state,
          values: {
            ...values as object,
            [field]: val,
          },
        }));
        // Validate against schema
        validateFormData<Values>(
          { ...values as any, [field]: val },
          validationSchema
        ).then(
          () => this.setState({ errors: {} }),
          (err: any) => this.setState({ errors: yupToFormErrors(err) })
        );
      };

      handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault();
        // setTouched();
        // setSubmitting(true);
        this.setState({
          touched: touchAllFields(this.state.values),
          isSubmitting: true,
        });
        const { values } = this.state;
        // Validate against schema
        validateFormData<Values>(values, validationSchema).then(
          () => {
            this.setState({ errors: {} });
            handleSubmit(mapValuesToPayload(this.state.values), {
              setTouched: this.setTouched,
              setErrors: this.setErrors,
              setError: this.setError,
              setSubmitting: this.setSubmitting,
              setValues: this.setValues,
              resetForm: this.resetForm,
              props: this.props,
            });
          },
          (err: any) =>
            this.setState({ isSubmitting: false, errors: yupToFormErrors(err) })
        );
      };

      handleBlur = (e: any) => {
        e.persist();
        const { name, id } = e.target;
        const field = name ? name : id;
        const { touched } = this.state;
        this.setTouched({ ...touched, [field]: true });
      };

      handleChangeValue = (field: string, value: any) => {
        const { touched, values } = this.state;
        // Set touched and form fields by name
        this.setState(state => ({
          ...state,
          values: {
            ...values as object,
            [field]: value,
          },
          touched: {
            ...touched as object,
            [field]: true,
          },
        }));
        // Validate against schema
        validateFormData<Values>(
          { ...values as any, [field]: value },
          validationSchema
        ).then(
          () => this.setState({ errors: {} }),
          (err: any) => this.setState({ errors: yupToFormErrors(err) })
        );
      };

      resetForm = (nextProps?: Props) => {
        this.setState({
          isSubmitting: false,
          errors: {},
          touched: {},
          error: undefined,
          values: nextProps
            ? mapPropsToValues(nextProps)
            : mapPropsToValues(this.props),
        });
      };

      handleReset = () => {
        this.setState({
          isSubmitting: false,
          errors: {},
          touched: {},
          error: undefined,
          values: mapPropsToValues(this.props),
        });
      };

      render() {
        return (
          <WrappedComponent
            {...this.props as any}
            {...this.state as any}
            setError={this.setError}
            setErrors={this.setErrors}
            setSubmitting={this.setSubmitting}
            setTouched={this.setTouched}
            setValues={this.setValues}
            resetForm={this.resetForm}
            handleReset={this.handleReset}
            handleSubmit={this.handleSubmit}
            handleChange={this.handleChange}
            handleBlur={this.handleBlur}
            handleChangeValue={this.handleChangeValue}
          />
        );
      }
    }
    // Make sure we preserve any custom statics on the original component.
    // @see https://github.com/apollographql/react-apollo/blob/master/src/graphql.tsx
    return hoistNonReactStatics(Formik, WrappedComponent);
  };
}
