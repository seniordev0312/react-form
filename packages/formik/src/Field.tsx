import * as React from 'react';
import invariant from 'tiny-warning';
import { useFormikContext } from './FormikContext';
import {
  FieldInputProps,
  FieldMetaProps,
  FormikProps,
  GenericFieldHTMLAttributes,
  SharedFieldProps,
} from './types';
import { useField, UseFieldProps } from './useField';
import { isEmptyChildren, isFunction } from './utils';

export interface FieldProps<V, FormValues> {
  field: FieldInputProps<V>;
  form: FormikProps<FormValues>; // if ppl want to restrict this for a given form, let them.
  meta: FieldMetaProps<V>;
}

export type FieldConfig<FieldValue, FormValues> = UseFieldProps<FieldValue> &
  SharedFieldProps<FieldProps<FieldValue, FormValues>>;

export function Field<FieldValue = any, FormValues = any>({
  render,
  children,
  as: is, // `as` is reserved in typescript lol
  component,
  ...props
}: GenericFieldHTMLAttributes & FieldConfig<FieldValue, FormValues>) {
  React.useEffect(() => {
    invariant(
      !render,
      `<Field render> has been deprecated. Please use a child callback function instead: <Field name={${name}}>{props => ...}</Field> instead.`
    );
    invariant(
      !(component && render),
      'You should not use <Field component> and <Field render> in the same <Field> component; <Field component> will be ignored'
    );

    invariant(
      !(is && children && isFunction(children)),
      'You should not use <Field as> and <Field children> as a function in the same <Field> component; <Field as> will be ignored.'
    );

    invariant(
      !(component && children && isFunction(children)),
      'You should not use <Field component> and <Field children> as a function in the same <Field> component; <Field component> will be ignored.'
    );

    invariant(
      !(render && children && !isEmptyChildren(children)),
      'You should not use <Field render> and <Field children> in the same <Field> component; <FastField children> will be ignored'
    );
    // eslint-disable-next-line
  }, []);

  const [field, meta] = useField<FieldValue>(props);
  const {
    validate: _validate,
    validationSchema: _validationSchema,
    ...formik
  } = useFormikContext<FormValues>();
  const legacyBag = { field, form: formik };

  if (render) {
    return render({ ...legacyBag, meta });
  }

  if (isFunction(children)) {
    return children({ ...legacyBag, meta });
  }

  if (component) {
    // This behavior is backwards compat with earlier Formik 0.9 to 1.x
    if (typeof component === 'string') {
      const { innerRef, parse, format, formatOnBlur, ...rest } = props;
      return React.createElement(
        component,
        { ref: innerRef, ...field, ...rest },
        children
      );
    }

    const { parse, format, formatOnBlur, ...rest } = props;
    return React.createElement(
      component,
      // @todo this is wrong type signature
      { ...legacyBag, meta, ...rest } as FieldProps<FieldValue, FormValues>,
      children
    );
  }

  // default to input here so we can check for both `as` and `children` above
  const asElement = is || 'input';

  if (typeof asElement === 'string') {
    const { innerRef, parse, format, formatOnBlur, ...rest } = props;
    return React.createElement(
      asElement,
      { ref: innerRef, ...field, ...rest },
      children
    );
  }
  const { parse, format, formatOnBlur, ...rest } = props;
  return React.createElement(
    asElement as React.ComponentType<FieldInputProps<FieldValue>>,
    { ...field, ...rest },
    children
  );
}
