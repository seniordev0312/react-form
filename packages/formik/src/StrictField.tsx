import * as React from 'react';
import { useContextSelector } from 'use-context-selector';
import { FormikContext } from './FormikContext';
import invariant from 'tiny-warning';
import {
  FieldInputProps,
  FieldMetaProps,
  FieldValidator,
  FormikContextType,
  GenericFieldHTMLAttributes,
} from './types';
import {
  defaultFormatFn,
  defaultParseFn,
  getIn,
  getValueFromEvent,
  isInputEvent,
  isObject,
  numberParseFn,
  isFunction,
  isEmptyChildren,
} from './utils';

export function unstable_useFieldValue(name: string) {
  return useContextSelector(FormikContext, ctx => getIn(ctx.values, name));
}

export function unstable_useFieldError(name: string) {
  return useContextSelector(FormikContext, ctx => getIn(ctx.errors, name));
}

export function unstable_useFieldTouched(name: string) {
  return useContextSelector(FormikContext, ctx => getIn(ctx.touched, name));
}

function useFieldState(name: string) {
  return {
    value: unstable_useFieldValue(name),
    touched: unstable_useFieldTouched(name),
    error: unstable_useFieldError(name),
    initialValue: useContextSelector(FormikContext, ctx =>
      getIn(ctx.initialValues.current, name)
    ),
    initialTouched: useContextSelector(
      FormikContext,
      ctx => !!getIn(ctx.initialTouched.current, name)
    ),
    initialError: useContextSelector(FormikContext, ctx =>
      getIn(ctx.initialErrors.current, name)
    ),
  };
}

export interface StrictFieldProps<V = any> {
  field: FieldInputProps<V>;
  meta: FieldMetaProps<V>;
}

export interface UseStrictFieldProps<V = any> {
  /**
   * Component to render. Can either be a string e.g. 'select', 'input', or 'textarea', or a component.
   */
  as?:
    | React.ComponentType<StrictFieldProps<V>['field']>
    | string
    | React.ComponentType
    | React.ForwardRefExoticComponent<any>;
  /**
   * Validate a single field value independently
   */
  validate?: FieldValidator;

  /**
   * Function to parse raw input value before setting it to state
   */
  parse?: (value: unknown, name: string) => V;

  /**
   * Function to transform value passed to input
   */
  format?: (value: V, name: string) => any;

  /**
   * Wait until blur event before formatting input value?
   * @default false
   */
  formatOnBlur?: boolean;

  /**
   * Field name
   */
  name: string;

  /** HTML input type */
  type?: string;

  /** Field value */
  value?: any;

  /** Inner ref */
  innerRef?: (instance: any) => void;
}

export type StrictFieldAttributes<T> = GenericFieldHTMLAttributes &
  UseStrictFieldProps<T> &
  T & { name: string };
export type StrictFieldHookConfig<T> = GenericFieldHTMLAttributes &
  UseStrictFieldProps<T>;
export type StrictFieldConfig<V = any> = UseStrictFieldProps<V> & {
  /**
   * Field component to render. Can either be a string like 'select' or a component.
   */
  component?:
    | string
    | React.ComponentType<StrictFieldProps<V>>
    | React.ComponentType
    | React.ForwardRefExoticComponent<any>;

  /**
   * Render prop (works like React router's <Route render={props =>} />)
   * @deprecated
   */
  render?: (props: StrictFieldProps<V>) => React.ReactNode;

  /**
   * Children render function <Field name>{props => ...}</Field>)
   */
  children?:
    | ((props: StrictFieldProps<V>) => React.ReactNode)
    | React.ReactNode;
};

export function unstable_useField<Value = any>(
  nameOrOptions: string | StrictFieldHookConfig<Value>
): [FieldInputProps<Value>, FieldMetaProps<Value>] {
  const isAnObject = isObject(nameOrOptions);
  // Normalize propsOrFieldName to FieldConfig<Value>
  const props: StrictFieldConfig<Value> = isAnObject
    ? (nameOrOptions as StrictFieldConfig<Value>)
    : ({ name: nameOrOptions as string } as StrictFieldConfig<Value>);

  const { name: fieldName, validate: validateFn } = props;

  const registerField = useContextSelector(
    FormikContext,
    (c: FormikContextType<unknown>) => c.registerField
  );

  const unregisterField = useContextSelector(
    FormikContext,
    (c: FormikContextType<unknown>) => c.unregisterField
  );

  React.useEffect(() => {
    if (fieldName) {
      registerField(fieldName, {
        validate: validateFn,
      });
    }
    return () => {
      if (fieldName) {
        unregisterField(fieldName);
      }
    };
  }, [registerField, unregisterField, fieldName, validateFn]);

  // const valueState = useFieldValue(fieldName);
  // const touchedState = useFieldTouched(fieldName);
  const meta = useFieldState(fieldName);
  const { value: valueState, touched: touchedState } = meta;
  const setFieldValue = useContextSelector(
    FormikContext,
    ctx => ctx.setFieldValue
  );
  const handleChange = useContextSelector(
    FormikContext,
    ctx => ctx.handleChange
  );
  const handleBlur = useContextSelector(FormikContext, ctx => ctx.handleBlur);

  const field: FieldInputProps<any> = {
    name: fieldName,
    value: valueState,
    // @todo extract into factory methods to use between formik and field
    onChange: handleChange,
    onBlur: handleBlur,
  };

  const {
    type,
    value: valueProp, // value is special for checkboxes
    as: is,
    multiple,
    parse = /number|range/.test(type ?? '') ? numberParseFn : defaultParseFn,
    format = defaultFormatFn,
    formatOnBlur = false,
  } = nameOrOptions as StrictFieldHookConfig<Value> & { multiple?: boolean }; // @todo why is this type failing?

  if (type === 'checkbox') {
    if (valueProp === undefined) {
      field.checked = !!valueState;
    } else {
      field.checked = !!(
        Array.isArray(valueState) && ~valueState.indexOf(valueProp)
      );
      field.value = valueProp;
    }
  } else if (type === 'radio') {
    field.checked = valueState === valueProp;
    field.value = valueProp;
  } else if (is === 'select' && multiple) {
    field.value = field.value || [];
    field.multiple = true;
  }

  if (type !== 'radio' && type !== 'checkbox' && !!format) {
    if (formatOnBlur === true) {
      if (touchedState === true) {
        field.value = format(field.value, fieldName);
      }
    } else {
      field.value = format(field.value, fieldName);
    }
  }

  // We incorporate the fact that we know the `name` prop by scoping `onChange`.
  // In addition, to support `parse` fn, we can't just re-use the OG `handleChange`, but
  // instead re-implement it's guts.
  if (type !== 'radio' && type !== 'checkbox') {
    field.onChange = (eventOrValue: React.ChangeEvent<any> | any) => {
      if (isInputEvent(eventOrValue)) {
        if (eventOrValue.persist) {
          eventOrValue.persist();
        }
        setFieldValue(
          fieldName,
          parse(getValueFromEvent(eventOrValue, valueState), fieldName)
        );
      } else {
        setFieldValue(fieldName, parse(eventOrValue, fieldName));
      }
    };
  }

  return [field, meta];
}

const useField = unstable_useField;

export function unstable_StrictField({
  render,
  children,
  as: is, // `as` is reserved in typescript lol
  component,
  ...props
}: StrictFieldAttributes<any>) {
  if (__DEV__) {
    // eslint-disable-next-line react-hooks/rules-of-hooks
    React.useEffect(() => {
      invariant(
        !render,
        `<Field render> has been deprecated and will be removed in future versions of Formik. Please use a child callback function instead. To get rid of this warning, replace <Field name="${name}" render={({field, form}) => ...} /> with <Field name="${name}">{({field, form, meta}) => ...}</Field>`
      );

      invariant(
        !(is && children && isFunction(children)),
        'You should not use <Field as> and <Field children> as a function in the same <Field> component; <Field as> will be ignored.'
      );

      invariant(
        !(component && children && isFunction(children)),
        'You should not use <Field component> and <Field children> as a function in the same <Field> component; <Field component> will be ignored.'
      );

      invariant(
        !(render && children && !isEmptyChildren(children)),
        'You should not use <Field render> and <Field children> in the same <Field> component; <Field children> will be ignored'
      );
      // eslint-disable-next-line
    }, []);
  }

  const [field, meta] = useField(props);

  if (render) {
    return render({ field, meta });
  }

  if (isFunction(children)) {
    return children({ field, meta });
  }

  const { innerRef, parse, format, formatOnBlur, validate, ...rest } = props;

  if (component) {
    // This behavior is backwards compat with earlier Formik 0.9 to 1.x
    if (typeof component === 'string') {
      return React.createElement(
        component,
        { ref: innerRef, ...field, ...rest },
        children
      );
    }
    return React.createElement(component, { field, meta, ...rest }, children);
  }

  // default to input here so we can check for both `as` and `children` above
  const asElement = is || 'input';

  if (typeof asElement === 'string') {
    return React.createElement(
      asElement,
      { ref: innerRef, ...field, ...rest },
      children
    );
  }
  return React.createElement(asElement, { ...field, ...rest }, children);
}
