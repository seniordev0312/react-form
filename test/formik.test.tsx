import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { Formik, InjectedFormikProps } from '../src/formik';
import { mount, shallow } from 'enzyme';

const Yup = require('yup');

describe('Formik', () => {
  interface Props {
    user: {
      name: string;
    };
    someFunction?: () => void;
  }

  interface Values {
    name: string;
  }

  const Form: React.SFC<InjectedFormikProps<Props, Values>> = ({
    values,
    handleSubmit,
    handleChange,
    handleBlur,
    errors,
    isSubmitting,
  }) => {
    return (
      <form onSubmit={handleSubmit}>
        <input
          type="text"
          onChange={handleChange}
          onBlur={handleBlur}
          value={values.name}
          name="name"
        />
        {errors.name &&
          <div id="feedback">
            {errors.name}
          </div>}
        {isSubmitting && <div id="submitting">Submitting</div>}
        <button type="submit">Submit</button>
      </form>
    );
  };

  it('renders Formik correctly', () => {
    const EnhancedForm = Formik<Props, Values, Values>({
      validationSchema: Yup.object().shape({
        name: Yup.string(),
      }),
      mapPropsToValues: ({ user }) => ({ ...user }),
      // tslint:disable-next-line:no-empty
      handleSubmit: () => {},
    })(Form);

    const tree = renderer
      .create(<EnhancedForm user={{ name: 'jared' }} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('initializes Formik state', () => {
    const EnhancedForm = Formik<Props, Values, Values>({
      validationSchema: Yup.object().shape({
        name: Yup.string(),
      }),
      mapPropsToValues: ({ user }) => ({ ...user }),
      // tslint:disable-next-line:no-empty
      handleSubmit: () => {},
    })(Form);
    const hoc = mount(<EnhancedForm user={{ name: 'jared' }} />);
    expect(hoc.state().isSubmitting).toBe(false);
    expect(hoc.state().touched).toEqual({});
    expect(hoc.state().values).toEqual({ name: 'jared' });
    expect(hoc.state().errors).toEqual({});
  });

  describe('handleSubmit', () => {
    it('calls preventDefault()', () => {
      const EnhancedForm = Formik<Props, Values, Values>({
        validationSchema: Yup.object().shape({
          name: Yup.string(),
        }),
        mapPropsToValues: ({ user }) => ({ ...user }),
        // tslint:disable-next-line:no-empty
        handleSubmit: () => {},
      })(Form);
      const hoc = mount(<EnhancedForm user={{ name: 'jared' }} />);
      const preventDefault = jest.fn();
      hoc.find('form').simulate('submit', { preventDefault });
      expect(preventDefault).toHaveBeenCalled();
    });

    it('touches all fields', () => {
      const EnhancedForm = Formik<Props, Values, Values>({
        validationSchema: Yup.object().shape({
          name: Yup.string(),
        }),
        mapPropsToValues: ({ user }) => ({ ...user }),
        // tslint:disable-next-line:no-empty
        handleSubmit: () => {},
      })(Form);
      const hoc = mount(<EnhancedForm user={{ name: 'jared' }} />);
      hoc.find('form').simulate('submit');
      expect(hoc.state().isSubmitting).toBe(true);
      expect(hoc.state().touched).toEqual({ name: true });
    });

    it('pushes submission state chagnes to child component', async () => {
      const EnhancedForm = Formik<Props, Values, Values>({
        validationSchema: Yup.object().shape({
          name: Yup.string().required(),
        }),
        mapPropsToValues: ({ user }) => ({ ...user }),
        handleSubmit: payload => {
          expect(payload).toEqual({ name: 'jared' });
        },
      })(Form);
      const hoc = shallow(<EnhancedForm user={{ name: 'jared' }} />);

      expect(hoc.find(Form).dive().find('#submitting')).toHaveLength(0);

      hoc.find(Form).dive().find('form').simulate('submit', {
        // tslint:disable-next-line:no-empty
        preventDefault() {},
      });
      expect(hoc.find(Form).dive().find('#submitting')).toHaveLength(1);
    });
  });

  describe('formikBag', () => {
    it('handleBlur sets touched', async () => {
      const EnhancedForm = Formik<Props, Values, Values>({
        validationSchema: Yup.object().shape({
          name: Yup.string().required(),
        }),
        mapPropsToValues: ({ user }) => ({ ...user }),
        handleSubmit: payload => {
          expect(payload).toEqual({ name: 'jared' });
        },
      })(Form);

      const hoc = shallow(<EnhancedForm user={{ name: 'jared' }} />);

      // Simulate a blur event in the inner Form component's input
      hoc.find(Form).dive().find('input').simulate('blur', {
        // tslint:disable-next-line:no-empty
        persist() {},
        target: {
          name: 'name',
        },
      });
      expect(hoc.update().state().touched).toEqual({ name: true });
    });

    it('handleChange sets values, and updates inputs', async () => {
      const EnhancedForm = Formik<Props, Values, Values>({
        validationSchema: Yup.object().shape({
          name: Yup.string().required(),
        }),
        mapPropsToValues: ({ user }) => ({ ...user }),
        handleSubmit: payload => {
          expect(payload).toEqual({ name: 'jared' });
        },
      })(Form);

      const hoc = shallow(<EnhancedForm user={{ name: 'jared' }} />);

      // Simulate a change event in the inner Form component's input
      hoc.find(Form).dive().find('input').simulate('change', {
        // tslint:disable-next-line:no-empty
        persist() {},
        target: {
          name: 'name',
          value: 'ian',
        },
      });

      expect(hoc.update().state().values).toEqual({ name: 'ian' });
      expect(
        hoc.update().find(Form).dive().find('input').props().value
      ).toEqual('ian');
    });
  });
});
