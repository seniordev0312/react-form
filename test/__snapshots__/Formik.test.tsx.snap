// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Formik renders Formik correctly 1`] = `
<form
  onSubmit={[Function]}
>
  <input
    name="name"
    onBlur={[Function]}
    onChange={[Function]}
    type="text"
    value="jared"
  />
  <button
    type="submit"
  >
    Submit
  </button>
</form>
`;
